### PR TITLE
Text box widget

### PR DESCRIPTION
**Pull request by QuadmasterXLII**
Several users (myself included) have wanted the ability to do text input as a widget. This text box widget is implemented by capturing keystrokes and adding them to a string, with special cases for keys like "enter", "left", "right" etc. 

http://stackoverflow.com/questions/25231449/do-text-number-input-fields-exist-in-matplotlib
http://stackoverflow.com/questions/25449398/matplotlib-user-enters-numeric-input-through-figure

### CODE REVIEW

**Comment by QuadmasterXLII** [comment]
It looks like I messed up some formatting. This is my first pull request, is the procedure to close this, fix the formatting, then open a new pull request?

**Comment by QuLogic** [comment]
No, you created a separate branch for this PR, so you can just add on to it and push. If it becomes too long, we might request that you rebase too.

**Comment by tacaswell** [comment]
There is at least one other quasi-dead PR open which tries to add this.  https://github.com/matplotlib/matplotlib/pull/1983

File: lib/matplotlib/widgets.py [hunk 1]
```diff
@@ -621,6 +622,247 @@ def disconnect(self, cid):
         except KeyError:
             pass
 
+class TextBox(AxesWidget):
+    """
+    A GUI neutral text input box.
+
+    For the text box to remain responsive
+    you must keep a reference to it.
```
Comment (mdboom) [inline lib/matplotlib/widgets.py]: Please make lines in docstrings as long as will fit in 80 columns.

File: lib/matplotlib/widgets.py [hunk 2]
```diff
@@ -621,6 +622,247 @@ def disconnect(self, cid):
...
+
+    The following attributes are accessible
```
Comment (mdboom) [inline lib/matplotlib/widgets.py]: colon at end of line

File: lib/matplotlib/widgets.py [hunk 3]
```diff
@@ -621,6 +622,247 @@ def disconnect(self, cid):
...
+      *label*
+        A :class:`matplotlib.text.Text` instance.
+
+      *color*
+        The color of the text box when not hovering.
+
+      *hovercolor*
+        The color of the text box when hovering.
+
+    Call :meth:`on_text_change` to be updated whenever the text changes
```
Comment (mdboom) [inline lib/matplotlib/widgets.py]: period at end of sentence.

File: lib/matplotlib/widgets.py [hunk 4]
```diff
@@ -621,6 +622,247 @@ def disconnect(self, cid):
...
+        
+        self.DIST_FROM_LEFT = .05        
+        
+        self.params_to_disable = []
+        for key in rcParams.keys():                   
+            if u'keymap' in key:
+               self.params_to_disable += [key]
+        
+        self.text = initial
+        self.label = ax.text(0.0,0.5, label,
```
Comment (mdboom) [inline lib/matplotlib/widgets.py]: PEP8: Run this file through the pep8 tool to catch things like spaces after punctuation, etc.

File: lib/matplotlib/widgets.py [hunk 5]
```diff
@@ -621,6 +622,247 @@ def disconnect(self, cid):
...
+        ax.set_yticks([])
+        self.color = color
+        self.hovercolor = hovercolor
+
+        self._lastcolor = color
+        
+        self.capturekeystrokes = False        
+           
+    def _make_text_disp(self, string):
+        return self.ax.text(self.DIST_FROM_LEFT, 0.5, string,
```
Comment (mdboom) [inline lib/matplotlib/widgets.py]: This spacing should be in display, not data coordinates.  That will be tricky and may require the use of the inverse transform, but will result in spacing that doesn't change with the size of the figure.

**Comment by mdboom** [comment]
This is cool work, and as you say, has been requested many times.

A couple of other things I'd like to see:  The mouse cursor should change to an I-beam when hovering over the text widget.  Clicking within the string should move the insertion cursor.

**Comment by QuadmasterXLII** [comment]
I went through and fixed the formatting issues you pointed out. Thanks! 

To my knowledge, there isn't currently a backend-agnostic way to set the cursor, so implementing the hover I-beam would require writing different code for each backend. All the other widgets light up when moused over, so I mimicked that behavior to indicate that this is a clickable element.

I will look into adding the ability to move the cursor by clicking, and moving to pixel coordinates for the cursor. 
Should I add kwargs passthorugh for the text objects?

**Comment by mdboom** [comment]
> To my knowledge, there isn't currently a backend-agnostic way to set the cursor, so implementing the hover I-beam would require writing different code for each backend. 

You can call `set_cursor` on a canvas, but I BEAM isn't currently one of the choices:

```
class Cursors(object):
    """Simple namespace for cursor reference"""
    HAND, POINTER, SELECT_REGION, MOVE = list(range(4))
cursors = Cursors()
```

It shouldn't be too much trouble to add that in all backends, though.  They all follow the same pattern.

> Should I add kwargs passthorugh for the text objects?

What do you mean, specifically?

**Comment by DanHickstein** [comment]
I've been wanting a textbox widget for a while, and I'm excited to see that there is already a PR in place! 

Any chance that this will move forward soon?

**Comment by QuadmasterXLII** [comment]
Yeah, it's good to hear that there's interest! I'll work on it this
weekend.

**Comment by DanHickstein** [comment]
One other small issue: the label text seems a little bit too close to the textbox, and adding spaces to the string has no effect.

An example: 
```python
import matplotlib.pyplot as plt
from matplotlib.widgets import TextBox
TextBox(plt.axes([0.5, 0.5, .2, .2]),'This is the label    ')
TextBox(plt.axes([0.5, 0.2, .2, .2]),'This is the label')
plt.show()
```
image

Thanks again for making this feature. I am using it make a quick-and-easy MPL-widgets based GUI and it's working nicely.

**Comment by QuadmasterXLII** [comment]
What should I do to fix the failed check "AppVeyor was unable to build non-mergable pull request" ?
Also, I have added the adjustable padding, and made the text box stop editing when the window is resized so that the cursor does not become misaligned.
I am working on the ability to move the cursor by clicking.
What else needs to be done for this to be pulled?

**Comment by QuLogic** [comment]
You will need to fetch master and [rebase](https://git-scm.com/docs/git-rebase) this branch to fix the merge conflicts.

**Comment by DanHickstein** [comment]
If you are using the [GitHub Desktop app](https://desktop.github.com/), then you just hit the button "Update from matplotlib/master". I think that the only conflict is with the ``CHANGELOG`` file. 

Is ``label_pad=0.01`` enough? I thought that 0.05 seemed prettier, but it probably depends on the figure size.

**Comment by tacaswell** [comment]
Pleas make sure you rebase + forcepush rather than merging master into your feature branch.

**Comment by DanHickstein** [comment]
Oh yes, I suppose that rebasing avoids a lot of clutter in the git history. 

Anyway, @QuadmasterXLII, you should go ahead and do that so that this can be merged.

**Comment by DanHickstein** [comment]
@QuadmasterXLII - you should rebase this so that it can be merged. I've been using these textboxes all the time for the last few months and they work great. (Well, pretty well. It's annoying that you can't click on move the cursor. But nothing's perfect...)

**Comment by QuadmasterXLII** [comment]
I did a rebase but appveyor s saying it fails. What should I do next?

**Comment by DanHickstein** [comment]
You can click on the appveyor build that failed and see the error. Unfortunately, it's not clear to me what's going wrong. It's a build from 18 days ago, so it hasn't been updated in a while. I'm not sure how to get it to rebuild aside from making some small commit to this fork.

**Comment by QuadmasterXLII** [comment]
I think it's ready for merger now!

**Comment by DanHickstein** [comment]
Nice work @QuadmasterXLII! 

It might also need a unit test in /lib/matplotlib/tests/test_widgets.py?

**Comment by Joshuaalbert** [comment]
You should add a set_val function, similar to the set_val function for Slider.
This can be done with

```
TextBox.text.text_disp('str')
plt.draw()
```

**Comment by fariza** [comment]
Really nice, one last thing @QuadmasterXLII can you squash the commits?

**Comment by Nodraak** [comment]
Hi!

I just tested this PR, it looks lovely, I don't know why this is not yet in matplotlib (yeah, nobody had taken care of this I guess :) ). I have just one small request:

When the user click outside of the text box, the `on_submit` callback is called. I believe this callback should be called only when the user actually validate its input with `Enter` or some `submit` button. As a workaround I have added `self.capturekeystrokes = False` just before [this](https://github.com/matplotlib/matplotlib/pull/5375/files#diff-5d314d27c09ac99e637cf11d0962aa97R862).

Anyway, thanks for this, it still get the work done !

**Comment by tacaswell** [comment]
@fariza Can you be responsible for getting this merged?  I am not super worried about squashing the commits, but the changes from @smithsp should probably get included (modulo the `set_axis_bgcolor` change).

We have not heard from @QuadmasterXLII , it might be most expedient if @fariza  just took over and opened a new PR.

@Nodraak Yes, development is mostly by volunteers with day jobs, things can fall through the cracks :disappointed: .  Could you work with @fariza and @QuadmasterXLII to get that change integrated with this PR?

**Comment by fariza** [comment]
Replaced by https://github.com/matplotlib/matplotlib/pull/6988

**Pull request closed without merging**